openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -459,46 +481,42 @@ def _on_user_authentication_failed(request):
         return HttpResponse(status=401)
 
     @staticmethod
-    def _verify_user(request, response, userid_in_session):
+    def _verify_user_and_log_mismatch(request, response, userid_in_session):
         """
-        Logs an error if the user marked at the time of process_request
-        does not match either the current user in the request or the
-        given userid_in_session.
+        Logs an error if the user has changed unexpectedly.
+
+        Other side effects:
+        - Sets a variety of custom attributes for unexpected user changes with
+            a 'safe_sessions.' prefix, like 'safe_sessions.session_id_changed'.
+        - May log additional details for users involved in a past unexpected user change,
+            if toggle is enabled. Uses the cache to track past user changes.
 
         Returns True if user matches in all places, False otherwise.
         """
-        # It's expected that a small number of views may change the
-        # user over the course of the request. We have exemptions for
-        # the user changing to/from None, but the login view can
-        # sometimes change the user from one value to another between
-        # the request and response phases, specifically when the login
-        # page is used during an active session.
-        #
-        # The relevant views set a flag to indicate the exemption.
-        if getattr(response, 'safe_sessions_expected_user_change', None):
-            return True
+        verify_user_results = SafeSessionMiddleware._verify_user_unchanged(request, response, userid_in_session)
+        if verify_user_results['verify_user'] is True:
+            # all is well; no unexpected user change was found
 
-        if not hasattr(request, 'safe_cookie_verified_user_id'):
-            # Skip verification if request didn't come in with a session cookie
-            return True
-
-        if hasattr(request.user, 'real_user'):
-            # If a view overrode the request.user with a masqueraded user, this will
-            #   revert/clean-up that change during response processing.
-            request.user = request.user.real_user
-
-        # determine if the request.user is different now than it was on the initial request
-        request_user_object_mismatch = request.safe_cookie_verified_user_id != request.user.id and\
-            request.user.id is not None
-
-        # determine if the current session user is different than the user in the initial request
-        session_user_mismatch = request.safe_cookie_verified_user_id != userid_in_session and\
-            userid_in_session is not None
+            try:
+                # possibly log request header if this user id was involved in an earlier mismatch
+                if LOG_REQUEST_USER_CHANGE_HEADERS.is_enabled() and request.user.id is not None:
+                    log_request_headers = cache.get(
+                        SafeSessionMiddleware._get_recent_user_change_cache_key(request.user.id), False
```

Comment:
I guess it doesn't matter _which_ user ID we do a cache check on, since we've only seen pairs.

Reply:
1. Since this is a request without a mismatch, we only have one user id to check.
2. When there is a mismatch, we add a cache entry for ALL user ids found during the mismatch.
So, I think we are in agreement that this should work.

Reply:
Oh right, this is the "normal" case.